//
// PROJECT:         Aspia
// FILE:            category/category_dmi_cache.proto
// LICENSE:         GNU Lesser General Public License 2.1
// PROGRAMMERS:     Dmitry Chapyshev (dmitry@aspia.ru)
//

syntax = "proto3";

option optimize_for = LITE_RUNTIME;

package aspia.proto;

message DmiCaches
{
    enum Location
    {
        LOCATION_UNKNOWN  = 0;
        LOCATION_INTERNAL = 1;
        LOCATION_EXTERNAL = 2;
        LOCATION_RESERVED = 3;
    }

    enum Status
    {
        STATUS_UNKNOWN  = 0;
        STATUS_ENABLED  = 1;
        STATUS_DISABLED = 2;
    }

    enum Mode
    {
        MODE_UNKNOWN                   = 0;
        MODE_WRITE_THRU                = 1;
        MODE_WRITE_BACK                = 2;
        MODE_WRITE_WITH_MEMORY_ADDRESS = 3;
    }

    enum SRAMType
    {
        SRAM_TYPE_BAD            = 0;
        SRAM_TYPE_OTHER          = 1;
        SRAM_TYPE_UNKNOWN        = 2;
        SRAM_TYPE_NON_BURST      = 4;
        SRAM_TYPE_BURST          = 8;
        SRAM_TYPE_PIPELINE_BURST = 16;
        SRAM_TYPE_SYNCHRONOUS    = 32;
        SRAM_TYPE_ASYNCHRONOUS   = 64;
    }

    enum ErrorCorrectionType
    {
        ERROR_CORRECTION_TYPE_UNKNOWN        = 0;
        ERROR_CORRECTION_TYPE_OTHER          = 1;
        ERROR_CORRECTION_TYPE_NONE           = 2;
        ERROR_CORRECTION_TYPE_PARITY         = 3;
        ERROR_CORRECTION_TYPE_SINGLE_BIT_ECC = 4;
        ERROR_CORRECTION_TYPE_MULTI_BIT_ECC  = 5;
    }

    enum Type
    {
        TYPE_UNKNOWN     = 0;
        TYPE_OTHER       = 1;
        TYPE_INSTRUCTION = 2;
        TYPE_DATA        = 3;
        TYPE_UNIFIED     = 4;
    }

    enum Associativity
    {
        ASSOCIATIVITY_UNKNOWN       = 0;
        ASSOCIATIVITY_OTHER         = 1;
        ASSOCIATIVITY_DIRECT_MAPPED = 2;
        ASSOCIATIVITY_2_WAY         = 3;
        ASSOCIATIVITY_4_WAY         = 4;
        ASSOCIATIVITY_FULLY         = 5;
        ASSOCIATIVITY_8_WAY         = 6;
        ASSOCIATIVITY_16_WAY        = 7;
        ASSOCIATIVITY_12_WAY        = 8;
        ASSOCIATIVITY_24_WAY        = 9;
        ASSOCIATIVITY_32_WAY        = 10;
        ASSOCIATIVITY_48_WAY        = 11;
        ASSOCIATIVITY_64_WAY        = 12;
        ASSOCIATIVITY_20_WAY        = 13;
    }

    message Item
    {
        string name                               = 1;
        Location location                         = 2;
        Status status                             = 3;
        Mode mode                                 = 4;
        int32 level                               = 5;
        int32 maximum_size                        = 6;
        int32 current_size                        = 7;
        uint32 supported_sram_types               = 8;
        SRAMType current_sram_type                = 9;
        int32 speed                               = 10;
        ErrorCorrectionType error_correction_type = 11;
        Type type                                 = 12;
        Associativity associativity               = 13;
    }

    repeated Item item = 1;
}
